scheds/rust/scx_mitosis/src/main.rs

```diff
@@ -0,0 +1,824 @@
+                reconfigured = true;
+                match self.reconfigure()? {
+                    Some(SplitOrMerge::Split(s)) => self.split(s)?,
+                    Some(SplitOrMerge::Merge(m)) => self.merge(m)?,
+                    _ => {
+                        trace!("No beneficial reconfiguration found");
+                        reconfigured = false;
+                    }
+                }
+                self.last_reconfiguration = std::time::Instant::now();
+            }
+            if reconfigured || self.last_cpu_rebalancing.elapsed() >= self.cpu_rebalancing_interval
+            {
+                trace!("CPU rebalancing");
+                self.assign_cpus(total_load)?;
+                self.last_cpu_rebalancing = std::time::Instant::now();
+                if reconfigured {
+                    self.update_cgroup_to_cell_assignment()?;
+                }
+                self.trigger_reconfiguration();
+            }
+        }
+        self.struct_ops.take();
+        uei_report!(&self.skel, uei)
+    }
+
+    fn update_cgroup_to_cell_assignment(&mut self) -> Result<()> {
+        for (cgroup, cell_idx) in self.cgroup_to_cell.iter() {
+            let mut cg_path = String::from("/sys/fs/cgroup/");
+            cg_path.push_str(cgroup);
+            let cg_file = File::open(&cg_path)
+                .with_context(|| format!("Failed to open path: {}", cg_path))?;
+            let cg_fd = cg_file.as_raw_fd();
+            let cg_fd_slice = unsafe { any_as_u8_slice(&cg_fd) };
+            let cell_idx_u32 = *cell_idx as libc::__u32;
+            let cell_idx_slice = unsafe { any_as_u8_slice(&cell_idx_u32) };
+            /* XXX: NO_EXIST should be correct here, but it fails */
+            self.skel
+                .maps()
+                .cgrp_cell_assignment()
+                .update(cg_fd_slice, cell_idx_slice, libbpf_rs::MapFlags::ANY)
+                .with_context(|| {
+                    format!("Failed to update cgroup cell assignment for: {}", cg_path)
+                })?;
+            trace!("Assigned {} to {}", cgroup, cell_idx);
+        }
+        self.skel.bss_mut().update_cell_assignment = true;
+        Ok(())
+    }
+
+    fn trigger_reconfiguration(&mut self) {
+        trace!("Triggering Reconfiguration");
+        self.skel.bss_mut().user_global_seq += 1;
+    }
+
+    /// Iterate through each cg in the cgroupfs, read its load from BPF and
+    /// update the cells map
+    fn collect_cgroup_load(&mut self) -> Result<f64> {
+        let mut cgroup_to_cell = HashMap::new();
+
+        for (_, cell) in self.cells.iter_mut() {
+            cell.cgroups = BTreeMap::new();
+            cell.load = 0.0;
+            cell.pinned_load = 0.0;
+        }
+        let now_mono = now_monotonic();
+        let mut stack = VecDeque::new();
+        let root = CgroupReader::root()?;
+        stack.push_back(root);
+        let mut total_load = 0.0;
+        while let Some(reader) = stack.pop_back() {
```

Comment:
Yeah, good point

Reply:
@htejun Hmm, thinking about this a bit more, the exact mechanics aren't clear to me. Here's my current idea: I could have BPF maintain a map of cg ID -> last_active_timestamp - updated on runnable/quiescent/cgroup_exit and periodically have the tick() remove any cg where last_active_timestamp > active_age_threshold (not sure how to synchronize tick() with the other BPF activity though). Then userspace can iterate the map and use the keys (cg ID) to get load #s.

Reply:
Yeah, it's a bit tricky because we don't want to make it expensive for the scheduling hot paths and creating low overhead interlocking is kinda challenging. One way could be sth like:
- There's global gen number and each cgroup has its own.
- There are two hashmaps.
- When a cgroup notices that its gen number differs from the current global, it updates its gen number and then puts its cgid in the even or odd hash map depending on the current gen number.
- When userspace wants to read active cgroups, it incs the gen number, waits for in flight updates to complete and then reads the hashtable.

There's the same problem where we don't have a way to RCU sync it, so the "wait for in flight updates" part can't be implemented correctly for now but otherwise it should work mostly okay? We can ask BPF folks about adding RCU constructs.

Reply:
Yeah, I thought about the two-hashmap solution, but this interacts oddly with wanting to read load at fairly high frequency - let's say I read every 100ms. I still would want to account for rebalancing cgroups not active in 100ms (in fact, that's the load half life). I guess I can keep track of all of that in userspace - maybe not so big of a deal.

Regarding RCU sync - does missing an in flight update matter?

Reply:
Yeah, the only thing the updated maps adds is a faster way of visiting what needs to be visited. After that, userspace logic should do what it needs to do. For the first draft, this all may just be a distraction but down the line I think this will become an actual problem.

re. RCU sync, yeah, it'll probably be okay for the most part. It's just not neat and can cause surprises down the line. It also makes it a bit difficult to reason about. RCU grace period can be pretty long too so it might not be a good fit anyway.

Reply:
Maybe userspace can always iterate both maps and clear only the inactive map? That'd guarantee that an active cgroup doesn't jump a period after an update. ie. A cgroup which was active at the tail end of period 1 showing up in period 2 is perfectly fine as the ordering isn't strictly defined anyway. However, it skipping period 2 and showing up in period 3 can lead to surprises. If both maps are always iterated from userspace, it will avoid such jumping and the added cost should be pretty small.